plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,1022 @@
+        """
+        # Distance from the source to the nearest gridpoint
+        dist = np.min(np.linalg.norm(self.grid_arr - self.source, axis=3))
+
+        # Find the particle with the highest speed towards the grid
+        vmax = np.max(np.dot(self.v, self.src_n))
+
+        # Time for fastest possible particle to reach the grid.
+        t = dist / vmax
+
+        # Coast the particles to the advanced position
+        self.x = self.x + self.v * t
+
+    def _generate_null_distribution(self):
+        r"""
+        Calculate the distribution of particles on the detector plane in the absence
+        of any simulated fields.
+
+        These positions are used to quickly compute null radiographs, which are
+        used to determine the degree of deflection.
+        """
+        # Calculate the unit vector from the source to the detector
+        dist = np.linalg.norm(self.src_det)
+        uvec = self.src_det / dist
+
+        # Calculate the remaining distance each particle needs to travel
+        # along that unit vector
+        remaining = np.dot(self.source, uvec)
+
+        # Calculate the time remaining to reach that plane and push
+        t = (dist - remaining) / np.dot(self.v, uvec)
+
+        # Calculate the particle positions for that case
+        self.x0 = self.source + self.v * np.outer(t, np.ones(3))
+
+    def _coast_to_detector(self):
+        r"""
+        Coasts all particles to the detector plane. This method will be
+        called after all particles have cleared the grid.
+
+        This step applies to all particles, including those that never touched
+        the grid.
+        """
+        dist_remaining = np.dot(self.x, self.det_n) + np.linalg.norm(self.detector)
+
+        v_towards_det = np.dot(self.v, -self.det_n)
+
+        # Time remaining for each particle to reach detector plane
+        t = dist_remaining / v_towards_det
+
+        # If particles have not yet reached the detector plane and are moving
+        # away from it, they will never reach the detector.
+        # So, we can remove them from the arrays
+        condition = np.logical_and(v_towards_det < 0, dist_remaining > 0)
+        ind = np.nonzero(np.where(condition, 0, 1))[0]
+        self.x = self.x[ind, :]
+        self.v = self.v[ind, :]
+        self.nparticles_grid = self.x.shape[0]
+        t = t[ind]
+
+        # Store the number of particles deflected
+        self.fract_deflected = (self.nparticles - ind.size) / self.nparticles
+
+        # Warn the user if a large number of particles are being deflected
+        if self.fract_deflected > 0.05:
+            warnings.warn(
+                f"{100*self.fract_deflected:.1f}% particles have been "
+                "deflected away from the detector plane. The fields "
+                "provided may be too high to successfully radiograph "
+                "with this particle energy.",
+                RuntimeWarning,
+            )
+
+        self.x += self.v * np.outer(t, np.ones(3))
```

Comment:
```suggestion
        self.x += self.v * t[:, np.newaxis]
```
Definitely don't do `outer` here, that incurs additional multiplications.

Reply:
Ooooooooooohhhh

I KNEW there had to be a better way to do this! Thanks!!